### PR TITLE
Allow for matching multiple resources with equal paths as in 4119

### PR DESCRIPTION
**Pull request by jansupol**
Signed-off-by: Jan Supol <jan.supol@oracle.com>

### CODE REVIEW

**Comment by jansupol** [comment]
fixes #4119

File: core-server/src/main/java/org/glassfish/jersey/server/internal/routing/PathMatchingRouter.java [hunk 1]
```diff
@@ -80,11 +92,42 @@
...
+
+    /**
+     * Return true iff the method is a sub-resource and not a locator and http method designator does not match
+     * the request http method designator
+     * @param route current route representing resource method / locator
+     * @param context Contains Request to check the http method
+     * @return false if method designator matches
+     */
+    private boolean designatorMismatch(final Route route, final RequestProcessingContext context) {
+        final String httpMethod = context.request().getMethod();
```
Comment (tomas-langer) [inline core-server/src/main/java/org/glassfish/jersey/server/internal/routing/PathMatchingRouter.java]: I would negate the method, so the if statement does not have to use `!`.
Also we can be more "clever" with the `HEAD` method - e.g. first try it for `HEAD` and then for `GET`, as `GET` can also respond to `HEAD` requests.

File: core-server/src/main/java/org/glassfish/jersey/server/model/RuntimeResource.java [hunk 2]
```diff
@@ -128,7 +128,16 @@ public RuntimeResource build(RuntimeResource parent) {
...
+            final int cmp = PathPattern.COMPARATOR.compare(o1.getPathPattern(), o2.getPathPattern());
+            int locatorCmp = 0;
+            return cmp != 0
+                    ? cmp
+                    // quaternary key sorting those derived from
+                    // sub-resource methods ahead of those derived from sub-resource locators
+                    : (locatorCmp = o1.resourceLocators.size() - o2.resourceLocators.size()) != 0
+                        ? locatorCmp
+                        // compare the regexes then
+                        : o2.regex.compareTo(o1.regex);
```
Comment (tomas-langer) [inline core-server/src/main/java/org/glassfish/jersey/server/model/RuntimeResource.java]: Nested ternary operators create quite unreadable code. I would refactor this to use an if statement instead.
```
final int cmp = PathPattern.COMPARATOR.compare(o1.getPathPattern(), o2.getPathPattern());
        if (cmp == 0) {
            final int locatorCmp = Integer.compare(o1.resourceLocators.size(), o2.resourceLocators.size());
            // quaternary key sorting those derived from
            // sub-resource methods ahead of those derived from sub-resource locators

            // compare the regexes if still equal
            return (locatorCmp == 0) ? o2.regex.compareTo(o1.regex) : locatorCmp;
        } else {
            return cmp;
        }
```

File: tests/e2e-server/src/test/java/org/glassfish/jersey/tests/e2e/server/routing/RegularExpressionsTest.java [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+public class RegularExpressionsTest extends JerseyTest {
+    public static String GET_VALUE = "get value";
+    public static String POST_VALUE = "post value";
+    public static String PUT_VALUE = "put value";
```
Comment (senivam) [inline tests/e2e-server/src/test/java/org/glassfish/jersey/tests/e2e/server/routing/RegularExpressionsTest.java]: final, possibly private?

File: core-server/src/main/java/org/glassfish/jersey/server/internal/routing/PathMatchingRouter.java [hunk 4]
```diff
@@ -80,11 +92,42 @@
...
+        final Router.Continuation result = Router.Continuation.of(context, acceptedRoute.next());
+
+        // tracing
+        tracingLogger.log(ServerTraceEvent.MATCH_PATH_SELECTED, acceptedRoute.routingPattern().getRegex());
+
+        return result;
+    }
+
+    /**
+     * Return true iff the method is a sub-resource and not a locator and http method designator does not match
```
Comment (senivam) [inline core-server/src/main/java/org/glassfish/jersey/server/internal/routing/PathMatchingRouter.java]: typo... (iff)

Comment (jansupol) [reply core-server/src/main/java/org/glassfish/jersey/server/internal/routing/PathMatchingRouter.java]: iff stands for [if and only if](https://en.wikipedia.org/wiki/If_and_only_if)

**Pull request merged**